idaes/generic_models/costing/costing_base.py

```diff
@@ -0,0 +1,598 @@
+                    f"Unit model {unit_model.name} already has a costing block"
+                    f" registered: {b.name}. Each unit may only have a single "
+                    "UnitModelCostingBlock associated with it.")
+        # Add block to unit model initialization order
+        unit_model._initialization_order.append(self)
+
+        # Register unit model with this costing package
+        fcb._registered_unit_costing.append(self)
+
+        # Get costing method if not provided
+        method = self.config.costing_method
+        if method is None:
+            method = fcb._get_costing_method_for(unit_model)
+
+        # Assign obejct references for costing package and unit model
+        add_object_reference(self, "costing_package", fcb)
+        add_object_reference(self, "unit_model", unit_model)
+
+        # Call unit costing method
+        method(self, **self.config.costing_method_arguments)
+
+        # Check that costs are Vars and have lower bound of 0
+        cost_vars = DefaultCostingComponents
+        for v in cost_vars:
+            try:
+                cvar = getattr(self, v)
+                if not isinstance(cvar, pyo.Var):
+                    raise TypeError(
+                        f"{unit_model.name} {v} component must be a Var. "
+                        "Please check the costing package you are using to "
+                        "ensure that all costing components are declared as "
+                        "variables.")
+                elif cvar.lb is None or cvar.lb < 0:
+                    _log.warn(
+                        f"{unit_model.name} {v} component has a lower bound "
+                        "less than zero. Be aware that this may result in "
+                        "negative costs during optimization.")
+            except AttributeError:
+                pass
+
+    def initialize(self, *args, **kwargs):
+        """
+        Placeholder method for initialization
+        """
+        # TODO: Implement an initialization method
+        # TODO: Need to have a general pupose method (block triangularisation?)
+        # TODO: Should also allow registering custom methods
+
+        # Vars and Constraints
+        for c in DefaultCostingComponents:
+            if hasattr(self, c):
+                var = getattr(self, c)
+                cons = getattr(self, f"{c}_constraint")
+                calculate_variable_from_constraint(var, cons)
+
+    def del_costing(self):
+        # Need to clean up references before deletion
+
+        # Alias flowsheet costing block and unit model (parent)
+        fcb = self.config.flowsheet_costing_block
+
+        # Clean up references from fsb and unit_model
+        try:
+            fcb._registered_unit_costing.remove(self)
+        except ValueError:
+            # Assume this is OK
+            pass
+        try:
+            self.unit_model._initialization_order.remove(self)
+        except ValueError:
+            # Assume this is OK
```

Comment:
Is this a good assumption?  I think we might want to not suppress this exception (this kind of thing has burned me in the past).

Reply:
I think I fixed this.